### ISSUE TITLE
A way to apply transformations after the bundle has been generated

### ISSUE THREAD

**Issue by jakearchibald** [rollup/rollup#3595]
### Feature Use Case

Sometimes a file within the build needs to know details about the build as a whole. For instance, a service worker needs to know the final bundle filenames for the things it wants to cache.

### Feature Proposal

In the past I've just modified the service worker file in `generateBundle`, but this will break its source map. Also, it doesn't seem easy to access `resolveFileUrl` at this point.

I guess I'm looking for an API like:

```js
this.modifyBundleEntry(chunkInfo, { code, map });
```

Where:

- `chunkInfo` is a `ChunkInfo`
- `code` is a `string`
- `map` is a `string | SourceMap`

This would update the `ChunkInfo`'s source, but also merge the sourcemap.

`this.modifyBundleEntry` would only be callable in `generateBundle` (so, I dunno, maybe it should go somewhere else).

Also:

```js
this.getResolvedFileURLSource(…)
```

This would at least take a `ChunkInfo`, representing the chunk the file was referenced from, and `fileName` for the file being referenced. I guess it may need other params?

**Comment by guybedford** [comment]
Did you try using the renderChunk hook here? I believe it should be
possible by treating the service worker file as a module in input (even if
it has no imports to other files). Or is the metadata you need not
available in that hook?

**Comment by jakearchibald** [comment]
`renderChunk` only gives you info about that chunk, not other chunks.

**Comment by guybedford** [comment]
In theory it should still be possible to query the graph metadata from
within renderChunk.

Did you try exploring the ‘this’ methods in that context? I forget if
there’s a new API there for graph queries or not...

**Comment by lukastaegert** [comment]
You can only query modules but not chunks. Technically, it would be possible to know the file names at that point due to the way hashing is implemented, but I am actively working towards changing the hashing. On the other hand, once hashing is re-implemented, one could get a list of chunks with placeholders instead of hashes and use those, which would be replaced by the actual hashes post-renderChunk by Rollup. The advantage would be that the hash of the modified file would reflect the actual content even for the new files. But this is future talk, I hope I can finally manage to implement this in the next two months.

But for now, there is no non-hacky solution for this.

**Comment by jakearchibald** [comment]
I don't think this issue is high priority, it can definitely wait for the hashing fixes.

**Comment by eight04** [comment]
👍 

How about to have a `pluginContext.getChunkInfo` API?

I'm trying to support dynamic import in rollup-plugin-iife. The main idea is to let users define a script loader, then the plugin would convert
```js
import("./foo-xxx.js")
```
to
```js
scriptLoader([...DEPENDENCIES_FOR_FOO, "./foo-xxx.js"], document.currentScript.src).then(() => fooXXX)
```
The problem is how to obtain `DEPENDENCIES_FOR_FOO`. Currently, I can only access all chunks at `writeBundle` hook and replace the string in the file, which doesn't work well with the source map.

It will be better if we can obtain dependencies at `renderChunk` hook, or even `renderDynamicImport` hook (which doesn't seems to work because filename is not available yet).
